### ISSUE TITLE
Python 3.10 regression: injector.UnknownProvider: couldn't determine provider

### ISSUE THREAD

**Issue by leinardi** [alecthomas/injector#192]
Some users of Fedora 35 [reported](https://gitlab.com/leinardi/gwe/-/issues/162) that an app I developed doesn't work anymore after upgrading from Fedora 43.

After some investigation it seems related to the upgrade to Python 3.10. I was able to reproduce the same issue on Ubuntu 21.10 after installing Python 3.10 and running the app on a venv that uses it.

This is the relevant traceback:
```
Traceback (most recent call last):
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 658, in get_binding
    return self._get_binding(interface, only_this_binder=is_scope)
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 653, in _get_binding
    raise KeyError
KeyError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/leinardi/Workspace/gitlab/gwe/.pycharm/gwe", line 43, in <module>
    from gwe import __main__
  File "/home/leinardi/Workspace/gitlab/gwe/gwe/__main__.py", line 33, in <module>
    from gwe.model.current_fan_profile import CurrentFanProfile
  File "/home/leinardi/Workspace/gitlab/gwe/gwe/model/__init__.py", line 19, in <module>
    from gwe.model.overclock_profile import OverclockProfile
  File "/home/leinardi/Workspace/gitlab/gwe/gwe/model/overclock_profile.py", line 29, in <module>
    OVERCLOCK_PROFILE_CHANGED_SUBJECT = INJECTOR.get(OverclockProfileChangedSubject)
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 952, in get
    binding, binder = self.binder.get_binding(interface)
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 667, in get_binding
    binding = self.create_binding(interface)
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 582, in create_binding
    provider = self.provider_for(interface, to)
  File "/home/leinardi/Workspace/gitlab/gwe/venv/lib/python3.10/site-packages/injector/__init__.py", line 644, in provider_for
    raise UnknownProvider('couldn\'t determine provider for %r to %r' % (interface, to))
injector.UnknownProvider: couldn't determine provider for gwe.di.OverclockProfileChangedSubject to None
```
The [code](https://gitlab.com/leinardi/gwe/-/blob/release/gwe/model/overclock_profile.py#L29) that is triggering the issue is this one:
```python
OVERCLOCK_PROFILE_CHANGED_SUBJECT = INJECTOR.get(OverclockProfileChangedSubject)
```
where `OverclockProfileChangedSubject` [is defined](https://gitlab.com/leinardi/gwe/-/blob/release/gwe/di.py#L35) as:
```python
OverclockProfileChangedSubject = NewType('OverclockProfileChangedSubject', Subject)
```

This code worked fine for several versions of Python and version 3.10 seems to be the only one affected.

**Comment by leinardi** [comment]
Is there any know workaroud (beside downgrading Python) for this issue?

**Comment by leinardi** [comment]
@jstasiak any suggestion?

**Comment by marvic2409** [comment]
Yep this issue is also affecting me.

**Comment by jstasiak** [comment]
No suggestions, sorry, pull requests fixing this are welcome though.

PS. Please provide the exact Python and Injector versions where this happens, for clarity.

**Comment by gardotd426** [comment]
This is now also an issue on Arch Linux, as it's been moved to 3.10 as well. It's going to be an issue everywhere eventually.

**Comment by Lamieur** [comment]
> PS. Please provide the exact Python and Injector versions where this happens, for clarity.
On Fedora 35 I've tested:
- Python 3.10.0 and Injector 0.18.4 (latest release)
- Python 3.10.0 and current git master branch of Injector,
- Python 3.10.1 (currently in updates-testing) and current git of Injector.

All these with gwe's git.

Every combination gives the same result.

**Comment by d4mler** [comment]
Same problem on Arch.

**Comment by WickedLukas** [comment]
I found this issue with a similar error message and appearently someone also found a solution for it : https://github.com/alecthomas/injector/issues/175
Unfortunately I am not knowledgeable enough, but maybe someone else can review it.

**Comment by Lamieur** [comment]
No, #175 is unrelated. I had the same thought as you, so I when I was testing with git branch of injector (my previous comment), I have in fact rebuilt injector with https://github.com/alecthomas/injector/pull/188 - that didn't do a thing for our case. It wast faster to check than to think on what Injector does or what Subject is, but apparently it's not "generic", whatever that is! ;)

Anyways, this is completely different, a regression caused by something in Python 3.10 for something which worked before.

(or RxPY 3.2.0, as that was on 3.1.1 in Fedora 34. 3.1.1 doesn't build with Python 3.10 so I can't check :))

**Comment by gardotd426** [comment]
RxPY was on 3.2.0 on Arch before Arch moved to 3.10, and this issue wasn't present, so it's not RxPY.

**Comment by Lamieur** [comment]
Good to know, I wanted to confirm but couldn't easily, thanks :)

**Comment by leinardi** [comment]
@jstasiak please have a look at #193

**Comment by leinardi** [comment]
@jstasiak Thank you for merging the fix :+1: 

Is there an ETA for the next release that will include this fix?

**Comment by gardotd426** [comment]
@leinardi I will say that I installed `python-injector-git` from the AUR (I had to edit the PKGBUILD because the writer of it has it in a broken state but it was a quick fix) and after installing that, GWE works perfectly.

**Comment by AbendGithub** [comment]
Just did exactly the same as @gardotd426. Edited the PKGBUILD in order to solve the issue. GWE started to work again. The hotfix did it.

**Comment by gardotd426** [comment]
It's no longer needed, after I reported the issue, the maintainer updated the PKGBUILD so all is good now. No editing of PKGBUILDs necessary. So anyone not able to wait can just use `python-injector-git` from the AUR (if they're on an Arch-based distro). Beyond that, I'd say this is probably venturing into off-topic territory from here on out

**Comment by jstasiak** [comment]
> @jstasiak Thank you for merging the fix 👍
> 
> Is there an ETA for the next release that will include this fix?

No worries. It's just been released as version 0.19.0, enjoy.
